discovery/seata-discovery-etcd3/src/main/java/io/seata/discovery/registery/etcd/EtcdRegistryServiceImpl.java

```diff
@@ -0,0 +1,403 @@
+    private static ConcurrentMap<String, Set<Watch.Listener>> listenerMap = null;
+    private static ConcurrentMap<String, EtcdWatcher> watcherMap = null;
+    private static long leaseId = 0;
+    private EtcdLifeKeeper lifeKeeper = null;
+    private Future<Boolean> lifeKeeperFuture = null;
+
+
+    private EtcdRegistryServiceImpl() {
+    }
+
+    /**
+     * get etcd registry service instance
+     *
+     * @return instance
+     */
+    static EtcdRegistryServiceImpl getInstance() {
+        if (null == instance) {
+            synchronized (EtcdRegistryServiceImpl.class) {
+                if (null == instance) {
+                    clusterAddressMap = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
+                    listenerMap = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
+                    watcherMap = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
+                    instance = new EtcdRegistryServiceImpl();
+                }
+            }
+        }
+        return instance;
+    }
+
+
+    @Override
+    public void register(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        doRegister(address);
+    }
+
+    /**
+     * do registry
+     *
+     * @param address
+     */
+    private void doRegister(InetSocketAddress address) throws Exception {
+        PutOption putOption = PutOption.newBuilder().withLeaseId(getLeaseId()).build();
+        getClient().getKVClient().put(buildRegestryKey(address), buildRegistryValue(address), putOption);
+    }
+
+
+    @Override
+    public void unregister(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        doUnregister(address);
+    }
+
+    /**
+     * do unregister
+     *
+     * @param address
+     * @throws Exception
+     */
+    private void doUnregister(InetSocketAddress address) throws Exception {
+        getClient().getKVClient().delete(buildRegestryKey(address));
+    }
+
+    @Override
+    public void subscribe(String cluster, Watch.Listener listener) throws Exception {
+        listenerMap.putIfAbsent(cluster, new HashSet<>());
+        listenerMap.get(cluster).add(listener);
+        EtcdWatcher watcher = watcherMap.computeIfAbsent(cluster, w -> new EtcdWatcher(listener));
+        EXECUTOR_SERVICE.submit(watcher);
+    }
+
+    @Override
+    public void unsubscribe(String cluster, Watch.Listener listener) throws Exception {
+        Set<Watch.Listener> subscribeSet = listenerMap.get(cluster);
+        if (null != subscribeSet) {
+            Set<Watch.Listener> newSubscribeSet = new HashSet<>();
+            for (Watch.Listener eventListener : subscribeSet) {
+                if (!eventListener.equals(listener)) {
+                    newSubscribeSet.add(eventListener);
+                }
+            }
+            listenerMap.put(cluster, newSubscribeSet);
+        }
+        watcherMap.remove(cluster).stop();
```

Comment:
can not stop EtcdLifeKeeper#process？

Reply:
> can not stop EtcdLifeKeeper#process？

EtcdLifeKeeper is used to refresh the lease which binding to registry key.like the name,it's for keeping the instance in healthy.I think when calling unsubscribe method, this means that the client no longer needs to get any updates from the registry, but the Seata server should still be running, so EtcdLifeKeeper should not be stopped.